### PR TITLE
Typed `Expr::map`, `map_typed`

### PR DESCRIPTION
**Pull request by rben01**
I'm wondering what the consensus on this function is. It's a strongly typed version of `Expr::map` that infers the output schema from the closure passed (which returns a `ChunkedArray<T>` instead of a `Series`), so you don't need the second `output_type` argument and can't get the schema wrong. Also you don't need to write `.into_series()` in the function body :D

```rust
/// Apply a strongly typed function/closure once the logical plan get executed.
///
/// Unlike [`Expr::map`], the closure returns a
/// `PolarsResult<Option<ChunkedArray<T>>>`, _not_ a `PolarsResult<Option<Series>>`.
/// Because the closure's return type is available at compile time, this means the
/// schema is guaranteed to be correct, and there is no need for a second
/// `output_type` argument.
pub fn map_typed<F, T>(self, function: F) -> Self
where
    F: Fn(Series) -> PolarsResult<Option<ChunkedArray<T>>> + 'static + Send + Sync,
    T: PolarsDataType,
    ChunkedArray<T>: IntoSeries,
{
    self.map(
        move |series| function(series).map(|opt| opt.map(|ca| ca.into_series())),
        GetOutput::from_type(T::get_dtype()),
    )
}
```

Sample usage:

```rust
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let df = df! [
        "a" => ["x", "yy", "zzz"]
    ]
    .unwrap()
    .lazy();

    let df = df
        .with_column(
            col("a")
                .map_typed(|col| Ok(Some(col.utf8()?.str_lengths())))
                .alias("lengths"),
        )
        .with_column(
            col("a")
                .map_typed(|col| Ok(Some(col.utf8()?.to_uppercase())))
                .alias("upper"),
        );

    println!("{:?}", df.collect()?);

    Ok(())
}
```

```
shape: (3, 3)
┌─────┬─────────┬───────┐
│ a   ┆ lengths ┆ upper │
│ --- ┆ ---     ┆ ---   │
│ str ┆ u32     ┆ str   │
╞═════╪═════════╪═══════╡
│ x   ┆ 1       ┆ X     │
│ yy  ┆ 2       ┆ YY    │
│ zzz ┆ 3       ┆ ZZZ   │
└─────┴─────────┴───────┘
```

Thoughts? I almost feel that this should supplant the dynamic `map` — how often do people really need the dynamism in the return type of their mapping closure? Off the top of my head, I can't think of a single time my `output_type` has been anything other than a compile type constant (e.g., `GetOutput::from_type(DataType::Whatever)`).

### CODE REVIEW

**Comment by ritchie46** [comment]
I am not sure about this. I understand that the strict typedness of this is much easier to work with, but this doesn't work for logical types such as `Categorical`, `Date`, `Decimal` etc.

If we add this we should clearly document that.

**Comment by rben01** [comment]
I see, their `get_dtype()` returns `Unknown`, I assume because those types need more data than just the data type itself (time unit, categories, etc.) Is it possible to have more strongly typed versions of these `PolarsDataType`s then 🤔?

**Comment by rben01** [comment]
Oh, probably not because there's nowhere to store that data at the type level. So, document that it won't work for data types that carry extra info (i.e., that the `DataType` variant has fields), and check (or panic) if the output dtype is `Unknown`? And given this limitation it should not supplant `map`.

**Comment by rben01** [comment]
Just pushed a change that I think may be more palatable. Usage:

```rust
let df = df
    .with_column(
        col("a")
            .map(
                |col| Ok(Some(col.utf8()?.str_lengths())),
                GetOutput::infer_type(),
            )
            .alias("lengths"),
    )
    .with_column(
        col("a")
            .map(
                |col| Ok(Some(col.utf8()?.to_uppercase())),
                GetOutput::infer_type(),
            )
            .alias("upper"),
    );
```
Works as expected.

Trying to use `GetOutput::infer_type()` when the closure returns something other than a `ChunkedArray<T>` returns an error:
```rust
let df = df
    .with_column(
        col("a")
            .map(
                // returns a Series, which is dynamic, so this won't work
                |col| Ok(Some(col.utf8()?.str_lengths().into_series())),
                GetOutput::infer_type(),
            )
            .alias("lengths"),
    )
```
Error:
```
Error: ComputeError(ErrString("Output type inference is only enabled for
ChunkedArray<T>. Because \"Series\" is a more complex and/or dynamic type, the
output type cannot be inferred from it. Either return a `ChunkedArray<T>` or use
`GetOutput::from_type(output_type)` with an explicit `output_type` as the output
type."))
```

File: polars/polars-lazy/polars-plan/src/dsl/mod.rs [hunk 1]
```diff
@@ -505,11 +505,24 @@ impl Expr {
...
+            if let Ok(dt) = A::get_simple_dtype() {
+                output_type = GetOutput::from_type(dt);
+            }
+        }
+
+        let f = move |s: &mut [Series]| {
+            if will_infer {
+                A::get_simple_dtype()?;
+            }
+            function(std::mem::take(&mut s[0])).map(|opt| opt.map(|arr| arr.into_series()))
```
Comment (ritchie46) [inline polars/polars-lazy/polars-plan/src/dsl/mod.rs:524]: I really don't think we need something extra in core only for this function.

File: polars/polars-lazy/polars-plan/src/dsl/expr.rs [hunk 2]
```diff
@@ -183,6 +195,10 @@ impl GetOutput {
         Default::default()
     }
 
+    pub fn infer_type() -> Self {
```
Comment (ritchie46) [inline polars/polars-lazy/polars-plan/src/dsl/expr.rs:198]: This is against polars rationale. Polars doesn't infer types at runtime (execution time of plan), but needs to know them statically before the plan is executed. 

If we infer, other parts of the plan will fail.

**Comment by rben01** [comment]
I'm going to close this since it seems against the Polars ethos and unlikely to be implemented.

**Pull request closed without merging**